File: long_questions/DSA/Leetcode/golang/7_Binary_Search/704_Binary_Search.go
package main

import "fmt"

// 704. Binary Search
// Time: O(log N), Space: O(1)
func search(nums []int, target int) int {
	left, right := 0, len(nums)-1
	
	for left <= right {
		mid := left + (right-left)/2
		
		if nums[mid] == target {
			return mid
		} else if nums[mid] < target {
			left = mid + 1
		} else {
			right = mid - 1
		}
	}
	
	return -1
}

/*
=======================================
PATTERN RECOGNITION & INSIGHTS
=======================================

## 1. ALGORITHM PATTERN: Classic Binary Search
- **Left Pointer**: Start of search range
- **Right Pointer**: End of search range
- **Middle Element**: Compare with target to narrow search
- **Range Reduction**: Halve search space each iteration

## 2. PROBLEM CHARACTERISTICS
- **Sorted Array**: Input array is sorted in ascending order
- **Unique Elements**: No duplicate elements (though algorithm works with duplicates)
- **Index Search**: Find index of target element
- **Logarithmic Time**: O(log N) time complexity

## 3. SIMILAR PROBLEMS
- Search in Rotated Sorted Array (LeetCode 33)
- Find First and Last Position of Element (LeetCode 34)
- Search Insert Position (LeetCode 35)
- Sqrt(x) (LeetCode 69)

## 4. KEY OBSERVATIONS
- **Monotonic property**: Sorted array enables binary search
- **Divide and conquer**: Halve search space each iteration
- **Boundary conditions**: Handle empty array and single element
- **Integer overflow**: Use left + (right-left)/2 instead of (left+right)/2

## 5. VARIATIONS & EXTENSIONS
- **Search Insert Position**: Find where to insert target
- **Lower/Upper Bound**: Find first/last occurrence
- **Rotated Array**: Handle rotated sorted arrays
- **Multiple Targets**: Find all occurrences

## 6. INTERVIEW INSIGHTS
- Always clarify: "Is array sorted? Are there duplicates?"
- Edge cases: empty array, single element, target not found
- Space complexity: O(1) - only pointers and variables
- Time complexity: O(log N) - halving search space

## 7. COMMON MISTAKES
- Using (left+right)/2 causing integer overflow
- Wrong loop condition (should be <= not <)
- Off-by-one errors in pointer updates
- Not handling empty array properly
- Forgetting to return -1 when not found

## 8. OPTIMIZATION STRATEGIES
- **Current solution is optimal**: O(log N) time, O(1) space
- **Recursive version**: Can be implemented recursively but uses stack space
- **Template method**: Use consistent binary search template
- **Early termination**: Not applicable (need to verify absence)

## 9. EXECUTION VISUALIZATION

## 10. HUMAN LOGIC PHASE

### Mental Model & Intuition
**Think of it like finding a name in a phone book:**
- You have a phone book sorted alphabetically (sorted array)
- You want to find a specific person's name (target)
- Instead of reading from start to finish, open to the middle
- If the name comes before your target, search the first half
- If the name comes after your target, search the second half
- Repeat this process until you find the name or run out of pages

### Step-by-Step Human Reasoning

#### Phase 1: Problem Understanding
1. **Input**: Sorted array of integers and target value
2. **Goal**: Find index of target in array
3. **Output**: Index if found, -1 if not found
4. **Constraint**: Array is sorted in ascending order

#### Phase 2: Key Insight Recognition
- **"Divide and conquer"** → Can eliminate half the elements each step
- **"Monotonic property"** → Sorted array enables efficient search
- **"Middle comparison"** → Compare middle element to target
- **"Range reduction"** → Narrow search space based on comparison

#### Phase 3: Strategy Development
```
Human thought process:
"I'll use two pointers, left and right, to define my search range.
I'll look at the middle element and compare it to my target.
If the middle element is my target, I'm done!
If it's smaller, my target must be in the right half.
If it's larger, my target must be in the left half.
I'll repeat this until I find it or the range is empty."
```

#### Phase 4: Edge Case Handling
- **Empty array**: Return -1 (no elements to search)
- **Single element**: Check if it equals target
- **Target not found**: Return -1 when left > right
- **Integer overflow**: Use safe middle calculation

#### Phase 5: Algorithm Walkthrough (Human Perspective)
```
Starting with: nums=[-1,0,3,5,9,12], target=9

Human thinking:
"I'll search for 9 in the sorted array:

Initial range: left=0, right=5 (entire array)
Middle index: (0+5)/2 = 2, nums[2]=3
3 < 9, so target must be in right half
New range: left=3, right=5

Middle index: (3+5)/2 = 4, nums[4]=9
9 == 9, found it! Return index 4"
```

#### Phase 6: Intuition Validation
- **Why binary search works**: Sorted array allows elimination of half elements
- **Why O(log N) time**: Each iteration halves the search space
- **Why O(1) space**: Only need pointers, no extra data structures
- **Why safe middle calculation**: Prevents integer overflow

### Common Human Pitfalls & How to Avoid Them
1. **"Why not just linear search?"** → Binary search is much faster for large arrays
2. **"Should I use recursion?"** → Iterative is usually better (no stack overflow)
3. **"What about duplicates?"** → Algorithm still works, returns one occurrence
4. **"Can I use (left+right)/2?"** → Risk of overflow, use left + (right-left)/2

### Real-World Analogy
**Like finding a page number in a book:**
- You have a book with numbered pages (sorted array)
- You want to find a specific page number (target)
- Open to the middle of the book
- If your page is higher, look in the second half
- If your page is lower, look in the first half
- Keep dividing until you find your page

### Human-Readable Pseudocode
```
function binarySearch(array, target):
    left = 0
    right = length(array) - 1
    
    while left <= right:
        middle = left + (right - left) // 2
        
        if array[middle] == target:
            return middle
        else if array[middle] < target:
            left = middle + 1
        else:
            right = middle - 1
    
    return -1  // Target not found
```

### Execution Visualization

### Example: nums=[-1,0,3,5,9,12], target=9
```
Array: [-1][0][3][5][9][12]
Index:   0  1  2  3  4  5

Initial: left=0, right=5
mid = 0 + (5-0)/2 = 2, nums[2]=3
3 < 9 → search right half
left=3, right=5

Step 2: left=3, right=5
mid = 3 + (5-3)/2 = 4, nums[4]=9
9 == 9 → FOUND! Return 4
```

### Example: Target not found: nums=[-1,0,3,5,9,12], target=2
```
Initial: left=0, right=5
mid=2, nums[2]=3 > 2 → search left half
left=0, right=1

Step 2: left=0, right=1
mid=0, nums[0]=-1 < 2 → search right half
left=1, right=1

Step 3: left=1, right=1
mid=1, nums[1]=0 < 2 → search right half
left=2, right=1

left > right → NOT FOUND, return -1
```

### Key Visualization Points:
- **Range boundaries**: left and right pointers define search space
- **Middle calculation**: Safe formula prevents overflow
- **Range reduction**: Eliminate half the elements each iteration
- **Termination**: Stop when found or range is empty

### Memory Layout Visualization:
```
Array: [-1][0][3][5][9][12]
Index:   0  1  2  3  4  5
        ^           ^
     left=0      right=5
        mid=2 (value=3)
```

### Time Complexity Breakdown:
- **Each iteration**: O(1) - constant work
- **Number of iterations**: O(log N) - halving search space
- **Total time**: O(log N)
- **Space**: O(1) - only pointers and variables

### Alternative Approaches:

#### 1. Linear Search (O(N) time, O(1) space)
```go
func search(nums []int, target int) int {
    for i, num := range nums {
        if num == target {
            return i
        }
    }
    return -1
}
```
- **Pros**: Simple, works for unsorted arrays
- **Cons**: O(N) time, much slower for large arrays

#### 2. Recursive Binary Search (O(log N) time, O(log N) space)
```go
func search(nums []int, target int) int {
    return binarySearchHelper(nums, target, 0, len(nums)-1)
}

func binarySearchHelper(nums []int, target, left, right int) int {
    if left > right {
        return -1
    }
    
    mid := left + (right-left)/2
    if nums[mid] == target {
        return mid
    } else if nums[mid] < target {
        return binarySearchHelper(nums, target, mid+1, right)
    } else {
        return binarySearchHelper(nums, target, left, mid-1)
    }
}
```
- **Pros**: Elegant, same time complexity
- **Cons**: Uses O(log N) stack space, risk of stack overflow

#### 3. Standard Library (O(log N) time, O(1) space)
```go
import "sort"

func search(nums []int, target int) int {
    idx := sort.SearchInts(nums, target)
    if idx < len(nums) && nums[idx] == target {
        return idx
    }
    return -1
}
```
- **Pros**: Uses standard library, tested
- **Cons**: Less learning value for interviews

### Extensions for Interviews:
- **Search Insert Position**: Find where to insert target
- **Lower Bound**: Find first element >= target
- **Upper Bound**: Find first element > target
- **Multiple Occurrences**: Find range of all target positions
*/
func main() {
	// Test cases
	testCases := []struct {
		nums   []int
		target int
	}{
		{[]int{-1, 0, 3, 5, 9, 12}, 9},
		{[]int{-1, 0, 3, 5, 9, 12}, 2},
		{[]int{1, 2, 3, 4, 5}, 3},
		{[]int{1, 2, 3, 4, 5}, 6},
		{[]int{}, 1},
		{[]int{1}, 1},
		{[]int{1}, 0},
		{[]int{-10, -5, 0, 5, 10}, -5},
		{[]int{-10, -5, 0, 5, 10}, 0},
		{[]int{2, 4, 6, 8, 10}, 7},
	}
	
	for i, tc := range testCases {
		result := search(tc.nums, tc.target)
		fmt.Printf("Test Case %d: nums=%v, target=%d -> Index: %d\n", 
			i+1, tc.nums, tc.target, result)
	}
}

File: long_questions/DSA/Leetcode/java/7_Binary_Search/704_Binary_Search.java
import java.util.Arrays;

public class BinarySearch {
    
    // 704. Binary Search
    // Time: O(log N), Space: O(1)
    public static int search(int[] nums, int target) {
        int left = 0, right = nums.length - 1;
        
        while (left <= right) {
            int mid = left + (right - left) / 2;
            
            if (nums[mid] == target) {
                return mid;
            } else if (nums[mid] < target) {
                left = mid + 1;
            } else {
                right = mid - 1;
            }
        }
        
        return -1;
    }

    public static void main(String[] args) {
        // Test cases
        int[][] testArrays = {
            {-1, 0, 3, 5, 9, 12},
            {-1, 0, 3, 5, 9, 12},
            {1, 2, 3, 4, 5},
            {1, 2, 3, 4, 5},
            {},
            {1},
            {1},
            {-10, -5, 0, 5, 10},
            {-10, -5, 0, 5, 10},
            {2, 4, 6, 8, 10}
        };
        
        int[] targets = {9, 2, 3, 6, 1, 1, 0, -5, 0, 7};
        
        for (int i = 0; i < testArrays.length; i++) {
            int result = search(testArrays[i], targets[i]);
            System.out.printf("Test Case %d: nums=%s, target=%d -> Index: %d\n", 
                i + 1, Arrays.toString(testArrays[i]), targets[i], result);
        }
    }

/*
=======================================
PATTERN RECOGNITION & INSIGHTS
=======================================

## 1. ALGORITHM PATTERN: Binary Search with Range Tracking
- **Binary Search**: Divide and conquer approach
- **Range Tracking**: Maintain left and right boundaries
- **Mid Calculation**: Prevent overflow with left + (right-left)/2
- **Target Comparison**: Adjust search range based on comparison

## 2. PROBLEM CHARACTERISTICS
- **Sorted Array**: Input array is sorted in ascending order
- **Single Target**: Need to find one specific element
- **Index Return**: Return position or -1 if not found
- **No Duplicates**: Each element appears at most once

## 3. SIMILAR PROBLEMS
- Search Insert Position
- Find First and Last Position of Element
- Search in Rotated Sorted Array
- Find Minimum in Rotated Sorted Array

## 4. KEY OBSERVATIONS
- Sorted array enables binary search O(log N)
- Mid calculation prevents integer overflow
- Can eliminate half of remaining elements each step
- Left/right pointers track current search range

## 5. VARIATIONS & EXTENSIONS
- Find first/last occurrence (with duplicates)
- Search in descending sorted array
- Find closest element to target
- Multiple queries on same array

## 6. INTERVIEW INSIGHTS
- Clarify: "Is array sorted in ascending order?"
- Clarify: "Can array contain duplicates?"
- Edge cases: empty array, single element, target not found
- Why binary search is better than linear search

## 7. COMMON MISTAKES
- Integer overflow in mid calculation
- Off-by-one errors in loop condition
- Not handling empty array case
- Wrong comparison direction for ascending order

## 8. OPTIMIZATION STRATEGIES
- Current solution is optimal
- For multiple queries, array stays sorted
- For cache efficiency, consider branch prediction
- For very large arrays, consider interpolation search

## 9. EXECUTION VISUALIZATION

## 10. HUMAN LOGIC PHASE

### Mental Model & Intuition
**Think of it like finding a name in a phone book:**
- You have a phone book sorted alphabetically (sorted array)
- You want to find a specific person's name (target)
- Instead of reading every page (linear search), you open to the middle
- Based on whether your target comes before or after, you eliminate half
- You keep narrowing down until you find the exact page

### Step-by-Step Human Reasoning

#### Phase 1: Problem Understanding
1. **Input**: Sorted array and target value
2. **Goal**: Find index of target in array
3. **Output**: Index or -1 if not found

#### Phase 2: Key Insight Recognition
- **"Array is sorted!"** → This is the key insight
- **"Can eliminate half!"** → Binary search principle
- **"How to find middle?"** → left + (right-left)/2
- **"When to stop?"** → When left > right

#### Phase 3: Strategy Development
```
Human thought process:
"I'll maintain left and right pointers for my search range.
I'll repeatedly look at the middle element:
- If it's my target, I'm done!
- If target is smaller, search left half
- If target is larger, search right half
I'll continue until I find it or exhaust the search space."
```

#### Phase 4: Edge Case Handling
- **Empty array**: Return -1 immediately
- **Single element**: Check if it matches target
- **Target not found**: Will exit when left > right
- **Integer overflow**: Use left + (right-left)/2 formula

#### Phase 5: Algorithm Walkthrough (Human Perspective)
```
Starting with: [-1, 0, 3, 5, 9, 12], target = 9

Human thinking:
"Initialize: left = 0, right = 5

Step 1: mid = 0 + (5-0)/2 = 2
           nums[2] = 3, target = 9
           3 < 9, so target is in right half
           left = 3, right = 5

Step 2: mid = 3 + (5-3)/2 = 4
           nums[4] = 9, target = 9
           Found it! Return index 4"
```

#### Phase 6: Intuition Validation
- **Why it works**: Each step eliminates half the search space
- **Why it's efficient**: Logarithmic time complexity
- **Why it's correct**: Systematic elimination guarantees finding target if exists

### Common Human Pitfalls & How to Avoid Them
1. **"Why not linear search?"** → Too slow O(N)
2. **"What about mid calculation?"** → Prevent overflow with proper formula
3. **"When to stop?"** → When left > right (not left >= right)
4. **"What about duplicates?"** → Standard binary search assumes unique elements

### Real-World Analogy
**Like finding a book in a library:**
- You have books arranged by Dewey Decimal number (sorted array)
- You want to find a specific book (target)
- You go to the middle shelf and section
- If your book's number is higher, you go right
- If lower, you go left
- You keep narrowing down until you find the exact shelf

### Human-Readable Pseudocode
```
function binarySearch(sortedArray, target):
    left = 0
    right = sortedArray.length - 1
    
    while left <= right:
        mid = left + (right - left) / 2
        
        if sortedArray[mid] == target:
            return mid
        else if sortedArray[mid] < target:
            left = mid + 1
        else:
            right = mid - 1
    
    return -1
```

### Execution Visualization

### Example: [-1, 0, 3, 5, 9, 12], target = 9
```
Initial: nums = [-1, 0, 3, 5, 9, 12], target = 9
left = 0, right = 5

Step 1: mid = 0 + (5-0)/2 = 2
→ nums[2] = 3, target = 9
→ 3 < 9, search right half
→ left = 3, right = 5
Range: [3, 4, 5] → [5, 9, 12]

Step 2: mid = 3 + (5-3)/2 = 4
→ nums[4] = 9, target = 9
→ Found! Return index 4

Final: Return 4 ✓
```

### Key Visualization Points:
- **Search range** halves each iteration
- **Mid calculation** prevents overflow
- **Comparison** determines which half to eliminate
- **Convergence** happens when target found or range exhausted

### Memory Layout Visualization:
```
Array: [-1, 0, 3, 5, 9, 12]
Index:   0   1   2   3   4   5

Step 1: left=0, right=5, mid=2
         [---|---]
         0   1   2   3   4   5
              ↑
              mid=2 (value=3 < target=9)
              Search right: [3,4,5]

Step 2: left=3, right=5, mid=4
         [-----|-]
         0   1   2   3   4   5
                      ↑
                      mid=4 (value=9 = target=9)
                      Found! ✓
```

### Time Complexity Breakdown:
- **Each Iteration**: Eliminates half of remaining elements
- **Number of Iterations**: O(log N) where N is array length
- **Total Time**: O(log N)
- **Space**: O(1) - only left, right, mid variables
- **Optimal**: Cannot do better for sorted array search
*/
